contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -97,29 +103,41 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
   uint256 internal constant FORWARDER_METADATA_LENGTH = 45;
   uint256 internal constant SIGNATURE_LENGTH = 65;
 
-  function setConfig(uint32 donId, uint8 f, address[] calldata signers) external onlyOwner {
+  function setConfig(uint32 donId, uint32 configVersion, uint8 f, address[] calldata signers) external onlyOwner {
     if (f == 0) revert FaultToleranceMustBePositive();
     if (signers.length > MAX_ORACLES) revert ExcessSigners(signers.length, MAX_ORACLES);
     if (signers.length <= 3 * f) revert InsufficientSigners(signers.length, 3 * f + 1);
 
-    // TODO: how does setConfig handle expiration? e.g. if the signer set changes
+    bytes32 configId = keccak256(abi.encode(donId, configVersion));
 
     // remove any old signer addresses
-    for (uint256 i; i < s_configs[donId].signers.length; ++i) {
-      address signer = s_configs[donId].signers[i];
-      delete s_configs[donId]._positions[signer];
+    for (uint256 i; i < s_configs[configId].signers.length; ++i) {
+      address signer = s_configs[configId].signers[i];
+      delete s_configs[configId]._positions[signer];
     }
 
     // add new signer addresses
-    s_configs[donId].signers = signers;
+    s_configs[configId].signers = signers;
     for (uint256 i; i < signers.length; ++i) {
       // assign indices, detect duplicates
       address signer = signers[i];
-      if (s_configs[donId]._positions[signer] != 0) revert DuplicateSigner(signer);
-      s_configs[donId]._positions[signer] = uint8(i) + 1;
-      s_configs[donId].signers.push(signer);
+      if (s_configs[configId]._positions[signer] != 0) revert DuplicateSigner(signer);
```

Comment:
Minor gas optimization - we could loop through the calldata and detect duplicates before storing values.

Reply:
would that be more efficient? If there's a duplicate then the call reverts and no store actually ocurred